### DISCUSSION TITLE
How to enable all rules via CLI but then ignore some via config file?

### DISCUSSION THREAD

**Discussion by sueskind** [astral-sh/ruff#4237] [Q&A] [answered]
I would like to enable all rules by default on the command-line (i.e. the `--select ALL` option) but then selectively disable some rules via a config file (e.g. `ruff.toml`).

Example:
I have the following script `main.py`:
```python
print("hello world")

```

When I run `ruff check main.py --select ALL` I get the following errors:
```
main.py:1:1: T201 `print` found
main.py:1:1: D100 Missing docstring in public module
```

Now I want to allow both of them via the `ruff.toml`:
```toml
ignore = ["D100", "T201"]
```

But running the same command again does not give any different results.

How would one be able to achieve this?

**Comment by dhruvmanila** [comment, marked as answer]
As per the [documentation](https://beta.ruff.rs/docs/configuration/#rule-selection) of rule selection:

> CLI options are given higher priority than `pyproject.toml` options, and the current `pyproject.toml` file is given higher priority than any inherited `pyproject.toml` files.

So, this means that the selecting all options through the CLI take precedence over ignoring some of them in config file.

The solution is to switch how you're providing the options:
* In your `pyproject.toml`, use `select = ["ALL"]`
* Ignore rules through the command-line flag `--ignore`

**Reply by sueskind** [reply]
Right, thanks for the answer. The thing is that I want to have a general job running in a pipeline but control the specific rules in each project that uses said pipeline separately via config files. By default however, all rules should be active.

If there is no way to "invert" this precedence I guess I have to do some shenanigans with temp files during a pipeline job.

Maybe there is some combination to achieve this by using the `--extend-*` options?

**Reply by charliermarsh** [reply]
I can't think of a way to do this, the CLI options will always win out. Could you instead use `"ALL"` in the configuration file, and disable rules for specific projects via `per-file-ignores`?

**Reply by buhtz** [reply]
Wow. I am shocked. Activating `ALL` rules increased my error counter from 0 to over 7000 on a very small project.

I know there is the error code catalog in the docs. But is there a summarizing text (e.g. a blog post) about the different rule categories in ruff, which one are active by default, which one not and maybe which one of the by default inactive ones could be considered?

Additionally wouldn't it be a good idea to give (on request) debug output about all rules that are active and inactive at the current time and configuration?

**Reply by dhruvmanila** [reply]
> I know there is the error code catalog in the docs. But is there a summarizing text (e.g. a blog post) about the different rule categories in ruff, which one are active by default, which one not and maybe which one of the by default inactive ones could be considered?

The default configuration is here: https://github.com/astral-sh/ruff#configuration

You can use the `--statistics` flag to summarize the list of rules which were violated for the current run.
